src/DataAPI.jl

```diff
@@ -515,6 +515,113 @@ Throw an error if `x` does not support metadata deletion for column `col`.
+
+The `read` field indicates whether reading metadata with the [`dimmetadata`](@ref)
+and [`dimmetadatakeys`]](@ref) functions is supported.
+
+The `write` field indicates whether modifying metadata with the [`dimmetadata!`](@ref),
+[`deletemetadata!`](@ref), and [`emptymetadata!`](@ref) functions is supported.
+"""
+dimmetadatasupport(::Type, i::Int) = (read=false, write=false)
+
+"""
+    dimmetadata(x, dim::Int, key::AbstractString, [default]; style::Bool=false)
+
+Return metadata value associated with `x` for dimension `dim` and key `key`.
+Throw an error if `x` does not support reading metadata for dimension `dim` or `x`
+supports reading metadata, but does not have a mapping for dimension `dim` for `key`.
+
+If `style=true` return a tuple of metadata value and metadata style. Metadata
+style is an additional information about the kind of metadata that is stored for
+the `key`.
+
+$STYLE_INFO
+
+If `default` is passed then return it if `x` supports reading metadata and has
+dimension `dim` but mapping for `key` is missing.
+If `style=true` return `(default, :default)`.
+"""
+function dimmetadata end
+
+"""
+    dimmetadata(x, dim::Int; style::Bool=false)
+
+Return a dictionary of metadata corresponding to dimension `dim` of object `x`. If `dim`
+ is not passed return a collection mapping each dimension to its metadata so that
+`collection[dim][key] == dimmetadata(x, dim, key)`.
+Throw an error if `x` does not support reading metadata corresponding to dimension `dim`.
+
+If `style=true` values are tuples of metadata value and metadata style. Metadata
+style is an additional information about the kind of metadata that is stored for
+the `key`.
+
+$STYLE_INFO
+
+The returned dictionary may be freshly allocated on each call to `dimmetadata` and
+is considered to be owned by `x` so it must not be modified.
+"""
+function dimmetadata(x::T, dim::Int; style::Bool=false) where {T}
+    if !dimmetadatasupport(T, dim).read
+        throw(ArgumentError("Objects of type $T do not support reading dimension metadata"))
+    end
+    return Dict(key => dimmetadata(x, dim, key, style=style) for key in dimmetadatakeys(x, dim))
+end
+function dimmetadata(x::T; style::Bool=false) where {T}
+    Tuple(dimmetadata(x, dim; style=style) for dim in 1:ndims(x))
+end
+
+"""
+    dimmetadatakeys(x, [dim::Int])
+
+If `dim` is passed return an iterator of metadata keys for which
+`dimmetadata(x, dim, key)` returns a metadata value. Throw an error if `x` does not
+support reading dimension metadata or if `dim` is not a dimension of `x`.
+
+If `dim` is not passed return an iterator of `dim => dimmetadatakeys(x, dim)`
+pairs for all dimensions that have metadata.
+If `x` does not support dimension metadata return `()`.
+"""
+function dimmetadatakeys(x, dim::Int)
+    dimmetadatasupport(typeof(x), dim).read || return ()
+    throw(MethodError(dimmetadatakeys, (x, dim)))
+end
+function dimmetadatakeys(x)
+    Dict(dim => dimmetadatakeys(x, dim) for dim in 1:ndims(x))
+end
```

Comment:
I don't remember the discussions offhand, but for `colmetadatakeys` we decided not to provide any fallback definition, in particular not to return `()` when metadata is not supported. Better be consistent for `dimmetadatakeys`, otherwise the API will be confusing (or, again, also adapt `colmetadatakeys` if necessary).

Reply:
The reason I had the fall back is because the docs still say "If `x` does not support column metadata return `()`". Should the docs for both of these be changed then?

Reply:
That sentence is in the paragraph regarding the case when `col` is omitted. When it's passed, the docstring  says that we throw an error. We cannot change this for `colmetadatakeys` anyway as it would be breaking, so better make `dimmetadatakeys` consistent with that.